File: src/test/java/no/fintlabs/repository/EventRepositoryInstanceFlowSummaryTest.java
package no.fintlabs.repository;

import lombok.extern.slf4j.Slf4j;
import no.fintlabs.repository.filters.InstanceFlowSummariesQueryFilter;
import no.fintlabs.repository.filters.InstanceFlowSummariesQueryFilter.InstanceFlowSummariesQueryFilterBuilder;
import no.fintlabs.repository.filters.InstanceStorageStatusQueryFilter;
import no.fintlabs.repository.filters.TimeQueryFilter;
import no.fintlabs.repository.projections.InstanceFlowSummaryProjection;
import no.fintlabs.repository.utils.instanceflowsummary.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.unit.DataSize;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Stream;

import static no.fintlabs.repository.utils.instanceflowsummary.Dataset.*;
import static no.fintlabs.repository.utils.instanceflowsummary.ParameterizedInstanceFlowTestCaseGenerator.createFilterPropertyTestCases;
import static org.assertj.core.api.Assertions.assertThat;

@Slf4j
@Testcontainers
@DataJpaTest(showSql = false)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@Transactional(propagation = Propagation.NOT_SUPPORTED)
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
public class EventRepositoryInstanceFlowSummaryTest {

    @Autowired
    EventRepository eventRepository;

    @Container
    static PostgreSQLContainer<?> postgreSQLContainer = new PostgreSQLContainer<>("postgres:16")
            .withUrlParam("reWriteBatchedInserts", "true")
            .withCreateContainerCmdModifier(createContainerCmd ->
                    Objects.requireNonNull(createContainerCmd.getHostConfig())
                            .withCpuCount(2L)
                            .withMemory(DataSize.ofGigabytes(8).toBytes())
            );

    @DynamicPropertySource
    static void postgreSQLProperties(DynamicPropertyRegistry registry) {
        postgreSQLContainer.start();
        registry.add("fint.database.url", postgreSQLContainer::getJdbcUrl);
        registry.add("fint.database.username", postgreSQLContainer::getUsername);
        registry.add("fint.database.password", postgreSQLContainer::getPassword);
    }

    private static boolean isInitialized = false;

    @BeforeEach
    public void setup() {
        if (isInitialized) {
            return;
        }
        isInitialized = true;
        eventRepository.saveAllAndFlush(
                ALL_EVENTS_AND_EXPECTED_SUMMARIES.stream()
                        .map(EventEntitiesAndExpectedSummary::getEventEntities)
                        .flatMap(Collection::stream)
                        .toList()
        );
    }

    public static Stream<Arguments> generateInstanceFlowSummariesTestCases() {
        FilterPropertyTestCases sourceApplicationIdTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::sourceApplicationIds,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(UNUSED_LONG_ID),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(1L),
                                Set.of(SA1_1_1, SA1_1_2)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(1L, 2L),
                                Set.of(SA1_1_1, SA1_1_2, SA2_2_1, SA2_3_3)
                        )
                )
        );
        FilterPropertyTestCases sourceApplicationIntegrationIdTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::sourceApplicationIntegrationIds,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(UNUSED_STRING_ID),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(SA_INTEGRATION_ID_1),
                                Set.of(SA1_1_1, SA1_1_2)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(SA_INTEGRATION_ID_3),
                                Set.of(SA2_3_3)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(SA_INTEGRATION_ID_1, SA_INTEGRATION_ID_2),
                                Set.of(SA1_1_1, SA1_1_2, SA2_2_1)
                        )
                )
        );
        FilterPropertyTestCases sourceApplicationInstanceIdTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::sourceApplicationInstanceIds,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(UNUSED_STRING_ID),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(SA_INSTANCE_ID_1),
                                Set.of(SA1_1_1, SA2_2_1)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(SA_INSTANCE_ID_4),
                                Set.of(SA3_4_4)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(SA_INSTANCE_ID_1, SA_INSTANCE_ID_4),
                                Set.of(SA1_1_1, SA2_2_1, SA3_4_4)
                        )
                )
        );
        FilterPropertyTestCases integrationIdTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::integrationIds,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(UNUSED_LONG_ID),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(101L),
                                Set.of(SA1_1_1, SA1_1_2)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(104L),
                                Set.of(SA3_4_4)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(101L, 104L),
                                Set.of(SA1_1_1, SA1_1_2, SA3_4_4)
                        )
                )
        );
        FilterPropertyTestCases timestampTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::timeQueryFilter,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMin(
                                        OffsetDateTime.of(2025, 1, 1, 0, 0, 0, 1000, ZoneOffset.UTC)
                                )
                                .build(),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMax(
                                        OffsetDateTime.of(2023, 1, 1, 0, 0, 0, 1000, ZoneOffset.UTC)
                                )
                                .build(),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMin(
                                        OffsetDateTime.of(2024, 1, 2, 8, 0, 0, 1000, ZoneOffset.UTC)
                                )
                                .build(),
                                Set.of(SA3_4_4)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMax(
                                        OffsetDateTime.of(2024, 1, 1, 11, 15, 0, 0, ZoneOffset.UTC)
                                )
                                .build(),
                                Set.of(SA2_2_1)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMin(
                                        OffsetDateTime.of(2024, 1, 1, 12, 0, 0, 0, ZoneOffset.UTC)
                                )
                                .latestStatusTimestampMax(
                                        OffsetDateTime.of(2024, 1, 2, 8, 0, 0, 2000, ZoneOffset.UTC)
                                )
                                .build(),
                                Set.of(SA1_1_1, SA1_1_2, SA3_4_4)
                        )
                )
        );
        FilterPropertyTestCases statusEventNamesTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::statusEventNames,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(UNUSED_STRING_ID),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(STATUS_EVENT_NAME_5),
                                Set.of(SA3_4_4)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(STATUS_EVENT_NAME_1, STATUS_EVENT_NAME_3),
                                Set.of(SA1_1_1, SA2_3_3)
                        )
                )
        );
        FilterPropertyTestCases storageStatusTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::instanceStorageStatusQueryFilter,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                new InstanceStorageStatusQueryFilter(
                                        List.of(STORAGE_STATUS_EVENT_NAME_1),
                                        null
                                ),
                                Set.of(SA1_1_2)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                new InstanceStorageStatusQueryFilter(
                                        List.of(STORAGE_STATUS_EVENT_NAME_1),
                                        null
                                ),
                                Set.of(SA1_1_2)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                new InstanceStorageStatusQueryFilter(
                                        null,
                                        true
                                ),
                                Set.of(SA1_1_1, SA2_2_1)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                new InstanceStorageStatusQueryFilter(
                                        List.of(STORAGE_STATUS_EVENT_NAME_1),
                                        true
                                ),
                                Set.of(SA1_1_1, SA1_1_2, SA2_2_1)
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                new InstanceStorageStatusQueryFilter(
                                        List.of(STORAGE_STATUS_EVENT_NAME_1, STORAGE_STATUS_EVENT_NAME_2),
                                        null
                                ),
                                Set.of(SA1_1_2, SA2_3_3, SA3_4_4)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                new InstanceStorageStatusQueryFilter(
                                        List.of(STORAGE_STATUS_EVENT_NAME_2),
                                        null
                                ),
                                Set.of(SA2_3_3, SA3_4_4)
                        )
                )
        );
        FilterPropertyTestCases destinationIdTestCases = createFilterPropertyTestCases(
                InstanceFlowSummariesQueryFilterBuilder::destinationIds,
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(UNUSED_STRING_ID),
                                Set.of()
                        ),
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(DESTINATION_INSTANCE_ID_1),
                                Set.of(SA2_3_3)
                        )
                ),
                List.of(
                        new FilterPropertyTestCaseConfiguration<>(
                                List.of(DESTINATION_INSTANCE_ID_1, DESTINATION_INSTANCE_ID_2),
                                Set.of(SA2_3_3, SA3_4_4)
                        )
                )
        );
        List<InstanceFlowSummariesTestCase> instanceFlowSummariesTestCases =
                ParameterizedInstanceFlowTestCaseGenerator.combineFilterPropertyTestCases(List.of(
                        sourceApplicationIdTestCases,
                        sourceApplicationIntegrationIdTestCases,
                        sourceApplicationInstanceIdTestCases,
                        integrationIdTestCases,
                        timestampTestCases,
                        statusEventNamesTestCases,
                        storageStatusTestCases,
                        destinationIdTestCases
                ));
        return instanceFlowSummariesTestCases
                .stream()
                .map(instanceFlowSummariesTestCase -> Arguments.of(
                        instanceFlowSummariesTestCase.getFilter(),
                        instanceFlowSummariesTestCase.getExpectedInstanceFlowSummaries()
                ));
    }

    @ParameterizedTest
    @MethodSource("generateInstanceFlowSummariesTestCases")
    public void instanceFlowSummariesParameterized(
            InstanceFlowSummariesQueryFilter filter,
            List<InstanceFlowSummaryProjection> expectedInstanceFlowSummaries
    ) {
        List<InstanceFlowSummaryProjection> instanceFlowSummaries = eventRepository.getInstanceFlowSummaries(
                filter,
                ALL_STATUS_EVENT_NAMES,
                ALL_STORAGE_STATUS_EVENT_NAMES,
                null
        );
        assertThat(instanceFlowSummaries).isEqualTo(expectedInstanceFlowSummaries);
    }
}

File: src/test/kotlin/no/novari/flyt/history/repository/EventRepositoryInstanceFlowSummaryTest.kt
package no.novari.flyt.history.repository

import no.novari.flyt.history.repository.filters.InstanceFlowSummariesQueryFilter
import no.novari.flyt.history.repository.filters.InstanceFlowSummariesQueryFilter.InstanceFlowSummariesQueryFilterBuilder
import no.novari.flyt.history.repository.filters.InstanceStorageStatusQueryFilter
import no.novari.flyt.history.repository.filters.TimeQueryFilter
import no.novari.flyt.history.repository.projections.InstanceFlowSummaryProjection
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.ALL_EVENTS_AND_EXPECTED_SUMMARIES
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.ALL_STATUS_EVENT_NAMES
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.ALL_STORAGE_STATUS_EVENT_NAMES
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.DESTINATION_INSTANCE_ID_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.DESTINATION_INSTANCE_ID_2
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA1_1_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA1_1_2
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA2_2_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA2_3_3
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA3_4_4
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA_INSTANCE_ID_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA_INSTANCE_ID_4
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA_INTEGRATION_ID_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA_INTEGRATION_ID_2
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.SA_INTEGRATION_ID_3
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.STATUS_EVENT_NAME_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.STATUS_EVENT_NAME_3
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.STATUS_EVENT_NAME_5
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.STORAGE_STATUS_EVENT_NAME_1
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.STORAGE_STATUS_EVENT_NAME_2
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.UNUSED_LONG_ID
import no.novari.flyt.history.repository.utils.instanceflowsummary.Dataset.Companion.UNUSED_STRING_ID
import no.novari.flyt.history.repository.utils.instanceflowsummary.FilterPropertyTestCaseConfiguration
import no.novari.flyt.history.repository.utils.instanceflowsummary.InstanceFlowSummariesTestCase
import no.novari.flyt.history.repository.utils.instanceflowsummary.ParameterizedInstanceFlowTestCaseGenerator
import no.novari.flyt.history.repository.utils.instanceflowsummary.ParameterizedInstanceFlowTestCaseGenerator.Companion.createFilterPropertyTestCases
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.unit.DataSize
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.stream.Stream

@Testcontainers(disabledWithoutDocker = true)
@DataJpaTest(showSql = false)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@Transactional(propagation = Propagation.NOT_SUPPORTED)
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
class EventRepositoryInstanceFlowSummaryTest {
    @Autowired
    lateinit var eventRepository: EventRepository

    @BeforeEach
    fun setup() {
        if (isInitialized) {
            return
        }
        isInitialized = true
        eventRepository.saveAllAndFlush(
            ALL_EVENTS_AND_EXPECTED_SUMMARIES.flatMap { it.eventEntities },
        )
    }

    @ParameterizedTest
    @MethodSource("generateInstanceFlowSummariesTestCases")
    fun instanceFlowSummariesParameterized(
        filter: InstanceFlowSummariesQueryFilter,
        expectedInstanceFlowSummaries: List<InstanceFlowSummaryProjection>,
    ) {
        try {
            val instanceFlowSummaries =
                eventRepository.getInstanceFlowSummaries(
                    filter,
                    ALL_STATUS_EVENT_NAMES,
                    ALL_STORAGE_STATUS_EVENT_NAMES,
                    null,
                )

            assertThat(instanceFlowSummaries).isEqualTo(expectedInstanceFlowSummaries)
        } catch (e: Exception) {
            e.printStackTrace()
            generateSequence<Throwable>(e) { it.cause }.forEachIndexed { index, throwable ->
                println("CAUSE[$index]: ${throwable::class.qualifiedName}: ${throwable.message}")
            }
            throw e
        }
    }

    @ParameterizedTest
    @MethodSource("generateInstanceFlowSummariesTestCases")
    fun instanceFlowSummariesTotalCountParameterized(
        filter: InstanceFlowSummariesQueryFilter,
        expectedInstanceFlowSummaries: List<InstanceFlowSummaryProjection>,
    ) {
        try {
            val instanceFlowSummariesTotalCount =
                eventRepository.getInstanceFlowSummariesTotalCount(
                    filter,
                    ALL_STATUS_EVENT_NAMES,
                    ALL_STORAGE_STATUS_EVENT_NAMES,
                )

            assertThat(instanceFlowSummariesTotalCount).isEqualTo(expectedInstanceFlowSummaries.size.toLong())
        } catch (e: Exception) {
            e.printStackTrace()
            generateSequence<Throwable>(e) { it.cause }.forEachIndexed { index, throwable ->
                println("CAUSE[$index]: ${throwable::class.qualifiedName}: ${throwable.message}")
            }
            throw e
        }
    }

    companion object {
        @JvmField
        @Container
        val postgreSQLContainer: PostgreSQLContainer<*> =
            PostgreSQLContainer("postgres:16")
                .withUrlParam("reWriteBatchedInserts", "true")
                .withCreateContainerCmdModifier { createContainerCmd ->
                    requireNotNull(createContainerCmd.hostConfig)
                        .withCpuCount(2L)
                        .withMemory(DataSize.ofGigabytes(8).toBytes())
                }

        private var isInitialized = false

        @JvmStatic
        @DynamicPropertySource
        fun postgreSQLProperties(registry: DynamicPropertyRegistry) {
            postgreSQLContainer.start()
            registry.add("fint.database.url", postgreSQLContainer::getJdbcUrl)
            registry.add("fint.database.username", postgreSQLContainer::getUsername)
            registry.add("fint.database.password", postgreSQLContainer::getPassword)
        }

        @JvmStatic
        fun generateInstanceFlowSummariesTestCases(): Stream<Arguments> {
            val sourceApplicationIdTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: List<Long>,
                        ->
                        builder.sourceApplicationIds(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(UNUSED_LONG_ID), emptySet()),
                        FilterPropertyTestCaseConfiguration(listOf(1L), setOf(SA1_1_1, SA1_1_2)),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(1L, 2L), setOf(SA1_1_1, SA1_1_2, SA2_2_1, SA2_3_3)),
                    ),
                )

            val sourceApplicationIntegrationIdTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: List<String>,
                        ->
                        builder.sourceApplicationIntegrationIds(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(UNUSED_STRING_ID), emptySet()),
                        FilterPropertyTestCaseConfiguration(listOf(SA_INTEGRATION_ID_1), setOf(SA1_1_1, SA1_1_2)),
                        FilterPropertyTestCaseConfiguration(listOf(SA_INTEGRATION_ID_3), setOf(SA2_3_3)),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            listOf(SA_INTEGRATION_ID_1, SA_INTEGRATION_ID_2),
                            setOf(SA1_1_1, SA1_1_2, SA2_2_1),
                        ),
                    ),
                )

            val sourceApplicationInstanceIdTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: List<String>,
                        ->
                        builder.sourceApplicationInstanceIds(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(UNUSED_STRING_ID), emptySet()),
                        FilterPropertyTestCaseConfiguration(listOf(SA_INSTANCE_ID_1), setOf(SA1_1_1, SA2_2_1)),
                        FilterPropertyTestCaseConfiguration(listOf(SA_INSTANCE_ID_4), setOf(SA3_4_4)),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            listOf(SA_INSTANCE_ID_1, SA_INSTANCE_ID_4),
                            setOf(SA1_1_1, SA2_2_1, SA3_4_4),
                        ),
                    ),
                )

            val integrationIdTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: List<Long>,
                        ->
                        builder.integrationIds(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(UNUSED_LONG_ID), emptySet()),
                        FilterPropertyTestCaseConfiguration(listOf(101L), setOf(SA1_1_1, SA1_1_2)),
                        FilterPropertyTestCaseConfiguration(listOf(104L), setOf(SA3_4_4)),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(101L, 104L), setOf(SA1_1_1, SA1_1_2, SA3_4_4)),
                    ),
                )

            val timestampTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: TimeQueryFilter,
                        ->
                        builder.timeQueryFilter(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMin(
                                    OffsetDateTime.of(2025, 1, 1, 0, 0, 0, 1000, ZoneOffset.UTC),
                                ).build(),
                            emptySet(),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMax(
                                    OffsetDateTime.of(2023, 1, 1, 0, 0, 0, 1000, ZoneOffset.UTC),
                                ).build(),
                            emptySet(),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMin(
                                    OffsetDateTime.of(2024, 1, 2, 8, 0, 0, 1000, ZoneOffset.UTC),
                                ).build(),
                            setOf(SA3_4_4),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMax(
                                    OffsetDateTime.of(2024, 1, 1, 11, 15, 0, 0, ZoneOffset.UTC),
                                ).build(),
                            setOf(SA2_2_1),
                        ),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            TimeQueryFilter
                                .builder()
                                .latestStatusTimestampMin(OffsetDateTime.of(2024, 1, 1, 12, 0, 0, 0, ZoneOffset.UTC))
                                .latestStatusTimestampMax(OffsetDateTime.of(2024, 1, 2, 8, 0, 0, 2000, ZoneOffset.UTC))
                                .build(),
                            setOf(SA1_1_1, SA1_1_2, SA3_4_4),
                        ),
                    ),
                )

            val statusEventNamesTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: List<String>,
                        ->
                        builder.statusEventNames(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(UNUSED_STRING_ID), emptySet()),
                        FilterPropertyTestCaseConfiguration(listOf(STATUS_EVENT_NAME_5), setOf(SA3_4_4)),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            listOf(STATUS_EVENT_NAME_1, STATUS_EVENT_NAME_3),
                            setOf(SA1_1_1, SA2_3_3),
                        ),
                    ),
                )

            val storageStatusTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: InstanceStorageStatusQueryFilter,
                        ->
                        builder.instanceStorageStatusQueryFilter(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            InstanceStorageStatusQueryFilter(listOf(STORAGE_STATUS_EVENT_NAME_1), null),
                            setOf(SA1_1_2),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            InstanceStorageStatusQueryFilter(listOf(STORAGE_STATUS_EVENT_NAME_1), null),
                            setOf(SA1_1_2),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            InstanceStorageStatusQueryFilter(null, true),
                            setOf(SA1_1_1, SA2_2_1),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            InstanceStorageStatusQueryFilter(listOf(STORAGE_STATUS_EVENT_NAME_1), true),
                            setOf(SA1_1_1, SA1_1_2, SA2_2_1),
                        ),
                        FilterPropertyTestCaseConfiguration(
                            InstanceStorageStatusQueryFilter(
                                listOf(STORAGE_STATUS_EVENT_NAME_1, STORAGE_STATUS_EVENT_NAME_2),
                                null,
                            ),
                            setOf(SA1_1_2, SA2_3_3, SA3_4_4),
                        ),
                    ),
                    listOf(
                        FilterPropertyTestCaseConfiguration(
                            InstanceStorageStatusQueryFilter(listOf(STORAGE_STATUS_EVENT_NAME_2), null),
                            setOf(SA2_3_3, SA3_4_4),
                        ),
                    ),
                )

            val destinationIdTestCases =
                createFilterPropertyTestCases(
                    {
                            builder: InstanceFlowSummariesQueryFilterBuilder,
                            value: List<String>,
                        ->
                        builder.destinationIds(value)
                    },
                    listOf(
                        FilterPropertyTestCaseConfiguration(listOf(UNUSED_STRING_ID), emptySet()),
                        FilterPropertyTestCaseConfiguration(
                            listOf(DESTINATION_INSTANCE_ID_1, DESTINATION_INSTANCE_ID_2),
                            setOf(SA2_3_3, SA3_4_4),
                        ),
                    ),
                    emptyList(),
                )

            val instanceFlowSummariesTestCases: List<InstanceFlowSummariesTestCase> =
                ParameterizedInstanceFlowTestCaseGenerator.combineFilterPropertyTestCases(
                    listOf(
                        sourceApplicationIdTestCases,
                        sourceApplicationIntegrationIdTestCases,
                        sourceApplicationInstanceIdTestCases,
                        integrationIdTestCases,
                        timestampTestCases,
                        statusEventNamesTestCases,
                        storageStatusTestCases,
                        destinationIdTestCases,
                    ),
                )

            return instanceFlowSummariesTestCases
                .stream()
                .map { instanceFlowSummariesTestCase ->
                    Arguments.of(
                        instanceFlowSummariesTestCase.filter,
                        instanceFlowSummariesTestCase.expectedInstanceFlowSummaries,
                    )
                }
        }
    }
}
